whatshap/variants.py

```diff
@@ -179,13 +202,60 @@ def _group_paired_reads(reads: Iterable[Read]) -> Iterator[List[Read]]:
         """
         groups = defaultdict(list)
         for read in reads:
-            groups[(read.source_id, read.name, read.sample_id)].append(read)
+            groups[(read.read.source_id, read.read.name, read.read.sample_id)].append(read)
+        n_skipped = 0
+        n_non_singleton = 0
         for group in groups.values():
             if len(group) > 2:
-                raise ReadSetError(
-                    f"Read name {group[0].name!r} occurs more than twice in the input file"
-                )
-            yield group
+                n_non_singleton += 1
+                logger.info(f"Group of read {group[0].read.name!r} has {len(group)} items.")
+                primary: Optional[ReadWithInfo] = None
+                for read in group:
+                    if not read.is_supplementary:
+                        if primary is not None:
+                            raise ReadSetError(
+                                f"Read name {group[0].name!r} has more than two primary alignments.")
+                        primary = read
+                if primary is None:
+                    n_skipped += 1
+                    continue
+                reference_start = primary.reference_start
```

Comment:
Can the code from here maybe be factored out into a separate function (`merge_reads` or so)? There is a bit much going on and the indentation is a bit too deep.